Get SQL cleanup working

rdlester opened the pull request:
Opening a pull request before all import scripts are updated to make sure everyone's okay with the general approach.

- Swtiched Mysql regex UDFs to a package I could actually get installed: https://github.com/mysqludf/lib_mysqludf_preg

- Added UDF installation to Dockerfile (testing this now).

- Abstracted shared cleanup commands into helper functions (data-imports/clean_utils.py - needs an underscore instead of a dash in the name in order to be imported in other scripts (?)). (tested)

- Updated import-dob-permits.py. (tested)

rdlester committed:
Extract SQL cleanup helper funcs

rdlester committed:
PCRE UDF installation in Docker

rdlester committed:
Use cleanup helpers in dob-permits script

jpfreeley commented:
It looks good to me.


On December 24, 2016 12:34:38 AM Ryan Lester <notifications@github.com> wrote:

> Opening a pull request before all import scripts are updated to make sure 
> everyone's okay with the general approach.
>
> - Swtiched Mysql regex UDFs to a package I could actually get installed: 
> https://github.com/mysqludf/lib_mysqludf_preg
>
> - Added UDF installation to Dockerfile (testing this now).
>
> - Abstracted shared cleanup commands into helper functions 
> (data-imports/clean_utils.py - needs an underscore instead of a dash in the 
> name in order to be imported in other scripts (?)). (tested)
>
> - Updated import-dob-permits.py. (tested)
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/chrishenry/deltanyc-heatseek/pull/35
>
> -- Commit Summary --
>
>   * Extract SQL cleanup helper funcs
>   * PCRE UDF installation in Docker
>   * Use cleanup helpers in dob-permits script
>
> -- File Changes --
>
>     M Dockerfile (8)
>     A data-imports/clean_utils.py (40)
>     M data-imports/import-dob-permits.py (58)
>
> -- Patch Links --
>
> https://github.com/chrishenry/deltanyc-heatseek/pull/35.patch
> https://github.com/chrishenry/deltanyc-heatseek/pull/35.diff
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/chrishenry/deltanyc-heatseek/pull/35

chrishenry commented on Dockerfile:
```diff
@@ -25,13 +25,19 @@ RUN wget --quiet https://github.com/krallin/tini/releases/download/v0.10.0/tini
+    mkdir lib_mysqludf_preg && \
+    unzip testing.zip -d lib_mysqludf_preg && \
+    cd lib_mysqludf_preg && \
+    ./configure && make install && make installdb && make test;
```
Hmm, I think this would need to be installed in the MySQL container.

Adding this will probably require us to build our own MySQL container. Or we could use [this](https://github.com/uqlibrary/docker-mysql-udf-preg)

rdlester replied:
Seems like the "this" will work fine. Tried adding it to the docker-compose.yml (pushed to this branch), but the DB failed to start up with a long list of errors, the most important bit of which I pasted below. Any ideas on what's wrong?

db_1   | 2016-12-24 16:51:34 7fa19625e720  InnoDB: Operating system error number 13 in a file operation.
db_1   | InnoDB: The error means mysqld does not have the access rights to
db_1   | InnoDB: the directory.
db_1   | 2016-12-24 16:51:34 13 [ERROR] InnoDB: Creating or opening ./ibdata1 failed!
db_1   | 2016-12-24 16:51:34 13 [ERROR] InnoDB: Could not open or create the system tablespace. If you tried to add new data files to the system tablespace, and it failed here, you should now edit innodb_data_file_path in my.cnf back to what it was, and remove the new ibdata files InnoDB created in this failed attempt. InnoDB only wrote those files full of zeros, but did not yet use them in any way. But be careful: do not remove old data files which contain your precious data!

chrishenry replied:
Ouch, those are file permission errors. As in, MySQL does not have permission to write to your datadir. This was pretty common with older versions of the MySQL container, and solved in a workable way earlier this year.

Looking at the [build history for the docker-mysql-udf-preg container](docker-mysql-udf-preg), the last build was a year ago, probably before the MySQL container worked out the kinks.

Options to move forward are, presented on a range of correct -> fast;

1) Poke the owner, see if he could push a new build. 
2) Fork the repo, and do it ourselves. (I'd be happy to do this)
3) Create a local Dockerfile for MySQL, copy-pasta the pertinent line, and have `docker-compose` manage the build for us.

chrishenry commented on data-imports/clean_utils.py:
```diff
@@ -0,0 +1,40 @@
+    """
+    # One line missing:
+    #    UPDATE {table} SET {column} = array_to_string(regexp_matches('(.*)(\d+)(?:TH|RD|ND|ST)( .+)'), '', {column} ) WHERE {column} ~ '.*(\d+)(?:TH|RD|ND|ST)( .+).*';
+    # No idea where regexp_matches is defined? Looks like it's a Postgresql command?
```
Yup, `regexp_matches` is a Postgres function, [documented here](https://www.postgresql.org/docs/9.1/static/functions-string.html)

rdlester committed:
Fix docker UDF setup?

rdlester commented:
My vote is number 2, and maybe send a pull request back to the original
repo. But, looking through the Dockerfile, I didn't see anything linked to
a particular version of the mysql container, so I'm not sure that's the
issue.

On Sat, Dec 24, 2016 at 12:39 PM Chris Henry <notifications@github.com>
wrote:

> *@chrishenry* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In Dockerfile <https://github.com/chrishenry/deltanyc-heatseek/pull/35>:
>
>
> > @@ -25,13 +25,19 @@ RUN wget --quiet https://github.com/krallin/tini/releases/download/v0.10.0/tini
>
>      mv tini /usr/local/bin/tini && \
>
>      chmod +x /usr/local/bin/tini
>
>
>
> +# Install Mysql regex UDFs
>
> +RUN wget --quiet https://github.com/mysqludf/lib_mysqludf_preg/archive/testing.zip && \
>
> +    mkdir lib_mysqludf_preg && \
>
> +    unzip testing.zip -d lib_mysqludf_preg && \
>
> +    cd lib_mysqludf_preg && \
>
> +    ./configure && make install && make installdb && make test;
>
>
>
> Ouch, those are file permission errors. As in, MySQL does not have
> permission to write to your datadir. This was pretty common with older
> versions of the MySQL container, and solved in a workable way earlier this
> year.
>
>
> Looking at the build history for the docker-mysql-udf-preg container
> <http://docker-mysql-udf-preg>, the last build was a year ago, probably
> before the MySQL container worked out the kinks.
>
>
> Options to move forward are, presented on a range of correct -> fast;
>
>
>
>
>
>    1. Poke the owner, see if he could push a new build.
>
>    2. Fork the repo, and do it ourselves. (I'd be happy to do this)
>
>    3. Create a local Dockerfile for MySQL, copy-pasta the pertinent line,
>    and have docker-compose manage the build for us.
>
>
>
>
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/chrishenry/deltanyc-heatseek/pull/35>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAmGQply0DwcPHMSx4kjFfPMZ50smGQnks5rLVjtgaJpZM4LVKhi>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

rdlester committed:
run_sql helper that runs and commits a multi SQL statement correctly

rdlester committed:
use cleanup helpers in 311 import

rdlester committed:
drop unused import from utils

rdlester commented:
Pushed some new updates:

- The code obtaining the cursor / connection was wrong; fixed and pulled out into a helper.

- Updated 311 import. Will keep updating later.

rdlester commented:
Looked more into the Docker stuff. One issue is that I wasn't pointing at the right place for the container (needed to use a git:// url, using the <name>/<repo> format points at docker.io?).

I forked the repo (https://github.com/rdlester/docker-mysql-udf-preg) and updated the install to build from HEAD but am still getting the same errors.

rdlester committed:
remove extracted code from dob-permits

rdlester committed:
new utils for boro replacements, indexes

rdlester committed:
update all import scripts

rdlester commented on data-imports/clean_utils.py:
```diff
@@ -0,0 +1,91 @@
+def clean_bbl(table, boro, block, lot):
+    return '''
+    SELECT concat(trim({boro}),trim(LPAD({block}, 5, '0')),
+            trim(LPAD({lot}, 4, '0'))) as bbl from hpd_buildings;
```
SELECT ... AS also appears to be Postgresql...

rdlester committed:
fix-ups; all imports tested

rdlester commented:
The bbl SQL code has been modified to something using standard SQL.

I've tested all the import scripts and they all seem to work. I haven't looked too closely at the output (I expect the address cleaning is still not good enough to link 100% of things together), and I do know there's a few cases that need additional cleaning (one table has additional boro codes "0" and "undefined"), but this is a good start.

Please take another look.

chrishenry commented:
👍
